src/lib/gsheets/index.js

```diff
@@ -0,0 +1,286 @@
+import { camelCase } from "lodash";
+import papa from "papaparse";
+
+import cacheFn from "@/promisetracker/lib/cache";
+import promisesQLFn from "@/promisetracker/lib/jsonql/promises";
+import theme from "@/promisetracker/theme";
+import { equalsIgnoreCase, formatDate, slugify } from "@/promisetracker/utils";
+
+function gsheets(server) {
+  const SPREADSHEETS_URL = "https://docs.google.com/spreadsheets/";
+  const spreadsheetId = server.env("GSHEETS_SPREADSHEET_ID");
+  const entitiesSheetId = server.env("GSHEETS_ENTITIES_SHEET_ID");
+  const sitesSheetId = server.env("GSHEETS_SITES_SHEET_ID");
+  const statusesSheetId = server.env("GSHEETS_STATUSES_SHEET_ID");
+  const categoriesSheetId = server.env("GSHEETS_CATEGORIES_SHEET_ID");
+  const promisesSheetId = server.env("GSHEETS_PROMISES_SHEET_ID");
+  const promisesCategoriesSheetId = server.env(
+    "GSHEETS_PROMISES_CATEGORIES_SHEET_ID"
+  );
+  const promisesEventsSheetId = server.env("GSHEETS_PROMISES_EVENTS_SHEET_ID");
+
+  const papaOptions = {
+    header: true,
+    transformHeader: (header) => camelCase(header.trim()),
+    transform: (value) => value?.trim(),
+  };
+
+  const papaPromise = (readableStream, options = papaOptions) => {
+    const parseStream = papa.parse(papa.NODE_STREAM_INPUT, options);
+    readableStream.pipe(parseStream);
+
+    return new Promise((resolve, reject) => {
+      const data = [];
+      parseStream.on("data", (chunk) => {
+        data.push(chunk);
+      });
+      parseStream.on("end", () => {
+        resolve(data);
+      });
+      parseStream.on("error", (error) => {
+        reject(error);
```

Comment:
I think this will fail on the first encounter of an error right? If so, wouldn't it be nice if we can just log the row with the error and continue with the rest?

Reply:
Yes it will fail @esirK ... On [node](https://github.com/mholt/PapaParse#papa-parse-for-node), the chunking/step options are actually not supported so it doesn't read row by row.